code/src/i2c_bus.cpp

```diff
@@ -0,0 +1,148 @@
+#include <i2c_bus.hpp>
+
+namespace r2d2::i2c {
+    void i2c_bus_c::pin_init() {
+        auto config_pin = [](uint32_t pin) {
+            PIOA->PIO_ABSR &= (~pin & PIOA->PIO_ABSR);
+            PIOA->PIO_PDR = pin;
+            PIOA->PIO_IDR = pin;
+            PIOA->PIO_PUER = pin;
+        };
+
+        if (_selected == TWI0) {
+            config_pin(PIO_PA18A_TWCK0);
+            config_pin(PIO_PA17A_TWD0);
+        } else if (_selected == TWI1) {
+            config_pin(PIO_PB13A_TWCK1);
+            config_pin(PIO_PB12A_TWD1);
+        }
+    }
+    void i2c_bus_c::clock_init() {
+        if (_selected == TWI0) {
+            if ((PMC->PMC_PCSR0 & (1 << ID_TWI0)) != (1 << ID_TWI0)) {
+                PMC->PMC_PCER0 |= 1 << ID_TWI0;
+            }
+        } else if (_selected == TWI1) {
+            if ((PMC->PMC_PCSR0 & (1 << ID_TWI1)) != (1 << ID_TWI1)) {
+                PMC->PMC_PCER0 |= 1 << ID_TWI1;
+            }
+        }
+
+        _selected->TWI_CR |= TWI_CR_SVEN;
+        _selected->TWI_CR |= TWI_CR_SWRST;
+        _selected->TWI_CR |= TWI_CR_SVDIS;
+        _selected->TWI_CR |= TWI_CR_MSDIS;
+        _selected->TWI_CR |= TWI_CR_MSEN;
+
+        static constexpr uint32_t masterClock =
```

Comment:
Why is this static?

Reply:
Why shouldn't it be? We aren't changing it afterwards

Reply:
Then only `constexpr` will suffice. It's very minor, but the compiler isn't allowed to omit `static constexpr` variables but it can omit `constexpr`. So technically this could prevent an optimization.